R/use_github_file.R

```diff
@@ -0,0 +1,130 @@
+#'   `path` will by copied from its current state in the repo's default branch.
+#' @inheritParams use_template
+#' @inheritParams use_github
+#'
+#' @return A logical indicator of whether a file was written, invisibly.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' use_github_file(
+#'   "https://github.com/r-lib/actions/blob/master/examples/check-standard.yaml"
+#' )
+#'
+#' use_github_file(
+#'   "r-lib/actions",
+#'   path = "examples/check-standard.yaml",
+#'   save_as = ".github/workflows/R-CMD-check.yaml"
+#' )
+#' }
+use_github_file <- function(repo_spec,
+                            path = NULL,
+                            save_as = NULL,
+                            ref = NULL,
+                            ignore = FALSE,
+                            open = FALSE,
+                            host = NULL) {
+
+  check_string(repo_spec)
+  dat <- parse_file_url(repo_spec)
+  if (dat$parsed) {
+    repo_spec <- dat$repo_spec
+    path      <- dat$path
+    ref       <- dat$ref
+    host      <- dat$host
+  }
+  check_string(path)
+  save_as <- save_as %||% path_file(path)
+  check_string(save_as)
+  maybe_string(ref)
+  maybe_string(host)
+
+  ref_string <- if (is.null(ref)) "" else glue("@{ref}")
+  github_string <- glue("{repo_spec}/{path}{ref_string}")
+  ui_done("Saving {ui_path(github_string)} to {ui_path(save_as)}")
+
+  # https://docs.github.com/en/rest/reference/repos#contents
+  # https://docs.github.com/en/rest/reference/repos#if-the-content-is-a-symlink
+  # If the requested {path} points to a symlink, and the symlink's target is a
+  # normal file in the repository, then the API responds with the content of the
+  # file....
+  tf <- withr::local_tempfile(
+    pattern = glue("use_github_file-{path_file(save_as)}-")
+  )
+  res <- gh::gh(
+    "/repos/{repo_spec}/contents/{path}",
+    repo_spec = repo_spec, path = path,
+    ref = ref,
+    .destfile = tf,
+    .accept = "application/vnd.github.v3.raw"
+  )
+
+  tf_contents <- read_utf8(tf)
+  new <- write_over(proj_path(save_as), tf_contents)
+
+  if (ignore) {
+    use_build_ignore(save_as)
+  }
+
+  if (open && new) {
+    edit_file(proj_path(save_as))
+  }
+
+  invisible(new)
+}
+
+# https://github.com/OWNER/REPO/blob/REF/path/to/some/file
+# https://raw.githubusercontent.com/OWNER/REPO/REF/path/to/some/file
+parse_file_url <- function(x) {
+  out <- list(
+    parsed = FALSE,
+    repo_spec = x,
+    path = NULL,
+    ref = NULL,
+    host = NULL
+  )
+
+  dat <- re_match(x, github_remote_regex)
+  if (is.na(dat$.match)) {
+    return(out)
+  }
+
+  if (!dat$host %in% c("raw.githubusercontent.com", "github.com") ||
+      !nzchar(dat$fragment) ||
+      (dat$host == "github.com" && !grepl("^/blob/", dat$fragment))) {
+    ui_stop("Can't parse the URL provided via {ui_code('repo_spec')}.")
+  }
+  out$parsed <- TRUE
+
+  dat$fragment <- sub("^/(blob/)?", "", dat$fragment)
+  dat_fragment <- re_match(dat$fragment, "^(?<ref>.+?)/(?<path>.+)$")
```

Comment:
I think this would be better written as something like

```r
"^(?<ref>[^/]+)/(?<path>.+)$"
```

In general I try to avoid lazy quantifiers when writing regular expressions, as they can cause catastrophic backtracking.